src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/HandlerResolver.cpp

```diff
@@ -208,83 +199,76 @@ HandlerResolver::FindNativeAssemblyFromGlobalLocation(
 //
 HRESULT
 HandlerResolver::FindNativeAssemblyFromHostfxr(
-    const HOSTFXR_OPTIONS& hostfxrOptions,
+    const HostFxrResolutionResult& hostfxrOptions,
     PCWSTR libraryName,
     std::wstring& handlerDllPath,
-    BaseOutputManager* outputManager
+    const IHttpApplication &pApplication,
+    const ShimOptions& pConfiguration,
+    std::shared_ptr<RedirectionOutput> stringRedirectionOutput
 )
+try
 {
     std::wstring   struNativeSearchPaths;
     size_t         intIndex = 0;
     size_t         intPrevIndex = 0;
     DWORD          dwBufferSize = s_initialGetNativeSearchDirectoriesBufferSize;
     DWORD          dwRequiredBufferSize = 0;
-    hostfxr_get_native_search_directories_fn pFnHostFxrSearchDirectories = nullptr;
 
     RETURN_LAST_ERROR_IF_NULL(m_hHostFxrDll = LoadLibraryW(hostfxrOptions.GetHostFxrLocation().c_str()));
 
-    try
-    {
-        pFnHostFxrSearchDirectories = ModuleHelpers::GetKnownProcAddress<hostfxr_get_native_search_directories_fn>(m_hHostFxrDll, "hostfxr_get_native_search_directories");
-    }
-    catch (...)
-    {
-        EventLog::Error(
-            ASPNETCORE_EVENT_GENERAL_ERROR,
-            ASPNETCORE_EVENT_HOSTFXR_DLL_INVALID_VERSION_MSG,
-            hostfxrOptions.GetHostFxrLocation().c_str()
-        );
-        return OBSERVE_CAUGHT_EXCEPTION();
-    }
-
-    RETURN_LAST_ERROR_IF_NULL(pFnHostFxrSearchDirectories);
-    struNativeSearchPaths.resize(dwBufferSize);
+    auto const hostFxr = HostFxr::CreateFromLoadedModule();
 
-    outputManager->TryStartRedirection();
-
-    while (TRUE)
     {
-        DWORD                       hostfxrArgc;
-        std::unique_ptr<PCWSTR[]>   hostfxrArgv;
-
-        hostfxrOptions.GetArguments(hostfxrArgc, hostfxrArgv);
-        const auto intHostFxrExitCode = pFnHostFxrSearchDirectories(
-            hostfxrArgc,
-            hostfxrArgv.get(),
-            struNativeSearchPaths.data(),
-            dwBufferSize,
-            &dwRequiredBufferSize
-        );
+        auto redirectionOutput = LoggingHelpers::CreateOutputs(
```

Comment:
From what I can tell, the flow of instantiating RedirectionOutput is:
- Create PipeOutputRedirection
- LoggingHelpers::CreateOutputs(pipeRedirection)
- Start stdOutRedirection
- Start hostFxrErrorRedirection

This seems overly complicated and easy to mess up. Few things I would consider:
- Why can't we create all 3 output redirections at the same time?
- Why can't the instantiation start redirection?

Here is the flow I would try to have:
- Call HostFxr::CreateLoadedModule before creating the RedirectionOutput.
- Call LoggingHelpers::CreateOutputs which would return an AggregateRedirectionOutput. Calling GetOutput can call into the StringStreamRedirectionOutput
- Redirection would start immediately on creation. Considering we already stop redirection in the destructor, this would make a nice parallel.

Reply:
> Why can't we create all 3 output redirections at the same time?

Because StringStream needs to be kept for much longer than file out console outputs for logging purposes.

> Why can't the instantiation start redirection?

No, hostfxr redirection has to be started on the thread that calls hostfxr method which is different from the one that creates outputs in the case of IN_PROCESS_APPLICATION.

> Call LoggingHelpers::CreateOutputs which would return an AggregateRedirectionOutput. Calling GetOutput can call into the StringStreamRedirectionOutput

Their scopes are very different.

> Redirection would start immediately on creation. Considering we already stop redirection in the destructor, this would make a nice parallel.

See point 2.

Reply:
I wish there was a cleaner way to do it though.